### ISSUE TITLE
Unexpected behaviour when using @JsonTypeInfo and custom @JsonTypeIdResolver

### ISSUE THREAD

**Issue by kamilzych** [FasterXML/jackson-databind#2811]
First of all, I am using the latest available version of Jackson, **2.11.2**
I'm trying to use a `@JsonTypeInfo` and a custom `@JsonTypeIdResolver` to deserialize an object which has a bounded type parameter field. Here are all classes (I removed getters and setters for clarity)

```java
interface Vehicle {
}

public class Car implements Vehicle {
    private int wheels;
    private String color;
}

public class Bicycle implements Vehicle {
    private int wheels;
    private String bicycleType;
}

public class Person<T extends Vehicle> {
    private String name;
    private Person.VehicleType vehicleType;

    @JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, include = JsonTypeInfo.As.EXTERNAL_PROPERTY, property = "vehicleType")
    @JsonTypeIdResolver(Person.VehicleTypeResolver.class)
    private T vehicle;

    public enum VehicleType {
        CAR(Car.class),
        BICYCLE(Bicycle.class);

        private final Class<? extends Vehicle> vehicleClass;

        VehicleType(Class<? extends Vehicle> vehicleClass) {
            this.vehicleClass = vehicleClass;
        }
    }

    private static class VehicleTypeResolver extends TypeIdResolverBase {

        private JavaType superType;

        @Override
        public void init(JavaType bt) {
            this.superType = bt;
        }

        public String idFromValue(Object value) {
            return null;
        }

        public String idFromValueAndType(Object value, Class<?> suggestedType) {
            return null;
        }

        @Override
        public JavaType typeFromId(DatabindContext context, String id) throws IOException {
            try {
                Class<? extends Vehicle> vehicleClass = Person.VehicleType.valueOf(id).vehicleClass;
                return context.constructSpecializedType(superType, vehicleClass);
            } catch (IllegalArgumentException e) {
                return null;
            }
        }

        public JsonTypeInfo.Id getMechanism() {
            return JsonTypeInfo.Id.NAME;
        }
    }
}

```
My `objectMapper` configuration looks like this:
```java
mapper = new ObjectMapper();
mapper.disable(DeserializationFeature.FAIL_ON_MISSING_EXTERNAL_TYPE_ID_PROPERTY);
mapper.disable(DeserializationFeature.FAIL_ON_INVALID_SUBTYPE);
mapper.enable(DeserializationFeature.READ_UNKNOWN_ENUM_VALUES_AS_NULL);
```
A valid JSON for happy case looks like this
```json
{
  "name": "kamil",
  "vehicleType": "CAR",
  "vehicle": {
    "wheels": 4,
    "color": "red"
  }
}
```
and this one works fine. What I want to achieve though is deserialization of JSON with missing `vehicleType` or `vehicle` so it doesn't throw an exception but puts `null` values to missing fields.

When the JSON is missing `vehicle` field or when invalid `vehicleType` is passed, it also works fine as those fields are indeed deserialized to `null` values. The only problem I have is a JSON with absent `vehicleType` in JSON.

```java
String json = "{ \"name\": \"kamil\", \"vehicle\": {\"wheels\": 4, \"color\": \"red\"}}";
Person<?> person = mapper.readValue(json, Person.class);
assertEquals("kamil", person.getName());
assertNull(person.getVehicleType());
assertNull(person.getVehicle());
```

this test doesn't pass as I get an exception
```
com.fasterxml.jackson.databind.exc.MismatchedInputException:
Unexpected token (END_OBJECT), expected VALUE_STRING: need JSON String that contains type id (for subtype of com.kamilzych.demo.model.Vehicle)
 at [Source: (String)"{ "name": "kamil", "vehicle": {"wheels": 4, "color": "red"}}"; line: 1, column: 60]
```

I created a demo project with this issue https://github.com/kamilzych/json-type-info-issue-demo it contains one test class with 4 tests, 1 of them (which I described above) does not pass. The demo project is using Java 8, it should be enough to run `mvn test`.

I'd appreciate taking a look. Either it's a bug or I'm doing something wrong. Thanks in advance.

**EDIT**
I also tried adding `defaultImpl = Void.class` to `@JsonTypeInfo` but that leads to `NullPointerException` in `TypeIdResolverBase` class.

**Comment by cowtowncoder** [comment]
Unfortunately I might not have time to look into this very soon, but wanted to add a note that I am aware of this report and will eventually try to figure it out.

**Comment by cowtowncoder** [comment]
Was able to reproduce as described, added failing test for 2.11.

**Comment by cowtowncoder** [comment]
Ok, the usage is not quite valid as shown. Problem is that if you have non-null data, there must be type for instantiating value, but without external type id there is no type to use (nominal base type is not generally safe is it is often abstract; Jackson will not speculate on if that might work).

So you will need to add `defaultImpl` with a real value, like:

    defaultImpl = Car.class

because using `Void.class` basically says "none" (which does and should lead to fail, although exception message is not optimal).
But in addition, if custom resolver is used, it also must implement `idFromValue()` and `idFromValueAndType()` methods (since they are used to get nominal type id for default case).
With those, test will work.

I did find that some of internal validation was incomplete, however, giving less useful failure information. I changed handling a bit to hopefully improve this.
